egressgateway: fix initial reconciliation

jibi opened the pull request:
When a new egress gateway manager is created, it will wait for the k8s
cache to be fully synced before running the first reconciliation.

Currently the logic is based on the WaitUntilK8sCacheIsSynced method
of the Daemon object, which waits on the k8sCachesSynced channel to be
closed (which indicates that the cache has been indeed synced).

The issue with this approach is that Daemon object is passed to
the NewEgressGatewayManager method _before_ its k8sCachesSynced
channel is properly initialized. This in turn causes the
WaitUntilK8sCacheIsSynced method to never return.

Since NewEgressGatewayManager must be called before that channel is
initialized, we need to switch to a polling approach, where the
k8sCachesSynced is checked periodically.

kkourt commented on pkg/egressgateway/manager.go:
```diff
@@ -57,7 +58,13 @@ func NewEgressGatewayManager(k8sCacheSyncedChecker k8sCacheSyncedChecker) *Manag
+			if manager.k8sCacheSyncedChecker.K8sCacheIsSynced() {
+				break
+			}
+
```
non-blocking comment: If there is a bug that does not allow the cache to be synced, we would get stuck in an infinite loop. Should we add an info message here? We can print it after N iterations if we think it would be too verbose.

jibi replied:
Looks like we are already logging this info in the `watchers` package: https://github.com/cilium/cilium/blob/25810847a11b659df9056a32b907e0da62fd5666/pkg/k8s/watchers/watcher.go#L391-L398

so probably no need to log it also here

kkourt reviewed (approved):
LGTM.
(Not a big fun of using Sleep, but I can't think of a better approach at this moment.)

jibi committed:
egressgateway: fix initial reconciliation

When a new egress gateway manager is created, it will wait for the k8s
cache to be fully synced before running the first reconciliation.

Currently the logic is based on the WaitUntilK8sCacheIsSynced method
of the Daemon object, which waits on the k8sCachesSynced channel to be
closed (which indicates that the cache has been indeed synced).

The issue with this approach is that Daemon object is passed to
the NewEgressGatewayManager method _before_ its k8sCachesSynced
channel is properly initialized. This in turn causes the
WaitUntilK8sCacheIsSynced method to never return.

Since NewEgressGatewayManager must be called before that channel is
initialized, we need to switch to a polling approach, where the
k8sCachesSynced is checked periodically.

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>
